pkg/kademlia/kademlia.go

```diff
@@ -0,0 +1,405 @@
+	p2p            p2p.Service           // p2p service to connect to nodes with
+	saturationFunc binSaturationFunc     // pluggable saturation function
+	connectedPeers *pslice.PSlice        // a slice of peers sorted and indexed by po, indexes kept in `bins`
+	knownPeers     *pslice.PSlice        // both are po aware slice of addresses
+	depth          uint8                 // current neighborhood depth
+	depthMu        sync.RWMutex          // protect depth changes
+	manageC        chan struct{}         // trigger the manage forever loop to connect to new peers
+	waitNext       map[string]time.Time  // sancation connections to a peer, key is overlay string and value is time to next retry
+	waitNextMu     sync.Mutex            // synchronize map
+	logger         logging.Logger        // logger
+	quit           chan struct{}         // quit channel
+	done           chan struct{}         // signal that `manage` has quit
+}
+
+// New returns a new Kademlia.
+func New(o Options) *Kad {
+	if o.SaturationFunc == nil {
+		o.SaturationFunc = binSaturated
+	}
+
+	k := &Kad{
+		base:           o.Base,
+		discovery:      o.Discovery,
+		addressBook:    o.AddressBook,
+		p2p:            o.P2P,
+		saturationFunc: o.SaturationFunc,
+		connectedPeers: pslice.New(maxBins),
+		knownPeers:     pslice.New(maxBins),
+		manageC:        make(chan struct{}, 1),
+		waitNext:       make(map[string]time.Time),
+		logger:         o.Logger,
+		quit:           make(chan struct{}),
+		done:           make(chan struct{}),
+	}
+
+	go k.manage()
+	return k
+}
+
+// manage is a forever loop that manages the connection to new peers
+// once they get added or once others leave.
+func (k *Kad) manage() {
+	var peerToRemove swarm.Address
+	defer close(k.done)
+
+	for {
+		select {
+		case <-k.quit:
+			return
+		case <-k.manageC:
+			err := k.knownPeers.EachBinRev(func(peer swarm.Address, po uint8) (bool, bool, error) {
+				if k.connectedPeers.Exists(peer) {
+					return false, false, nil
+				}
+
+				k.waitNextMu.Lock()
+				if next, ok := k.waitNext[peer.String()]; ok && time.Now().Before(next) {
+					k.waitNextMu.Unlock()
+					return false, false, nil
+				}
+				k.waitNextMu.Unlock()
+
+				currentDepth := k.NeighborhoodDepth()
+				if k.saturationFunc(po, currentDepth, k.connectedPeers) {
+					return false, false, nil
+				}
+
+				ma, err := k.addressBook.Get(peer)
+				if err != nil {
+					// either a peer is not known in the address book, in which case it
+					// should be removed, or that some severe I/O problem is at hand
+
+					if errors.Is(err, storage.ErrNotFound) {
+						k.logger.Errorf("failed to get address book entry for peer: %s", peer.String())
+						peerToRemove = peer
+						return false, false, errMissingAddressBookEntry
+					}
+
+					return false, false, err
+				}
+
+				k.logger.Debugf("kademlia dialing to peer %s", peer.String())
+				_, err = k.p2p.Connect(context.Background(), ma)
```

Comment:
actually, this is not _so_ trivial, since this approach you're highlighting would actually always leak a goroutine that would deadlock until the node shuts down. but yeah i get the point. i'll try to devise something

Reply:
Look at the pusher.go. It doesn't leak there. But i am glad you got the point.